src/hotspot/share/utilities/globalDefinitions.hpp

```diff
@@ -512,15 +513,20 @@ inline size_t pointer_delta(const MetaWord* left, const MetaWord* right) {
 // warnings, for example when truncating a size_t to an int when we
 // know the size_t is a small struct. Such casts are risky because
 // they effectively disable useful compiler warnings. We can make our
-// lives safer with this function, which ensures that any cast is
-// reversible without loss of information. It doesn't check
-// everything: it isn't intended to make sure that pointer types are
+// lives safer with this function, which ensures that any cast
+// fits within the size of the type cast to, and tolerates sign extension.
+// It doesn't check everything: it isn't intended to make sure that pointer types are
 // compatible, for example.
 template <typename T2, typename T1>
-T2 checked_cast(T1 thing) {
-  T2 result = static_cast<T2>(thing);
-  assert(static_cast<T1>(result) == thing, "must be");
-  return result;
+constexpr T2 checked_cast(T1 thing) {
+#ifdef ASSERT
+  typedef typename std::make_signed<T1>::type S1;
+  S1 test = static_cast<S1>(thing);
+  S1 max = std::numeric_limits<T2>::max();
+  S1 min = (std::is_signed<T2>::value) ? std::numeric_limits<T2>::min() : -max;
+  assert(test <= max && test >= min, "out of range of destination type");
+#endif
+  return static_cast<T2>(thing);
 }
```

Comment:
I don't understand this change.  Isn't it confusing to treat unsigned types as signed here?  Unsigned types don't sign extend.  If emit_intXX can take signed values then shouldn't that be reflected in the types of the arguements? 
Also, -max should be -max - 1 for 2's complement, right?

Reply:
The conversion from an unsigned type to a smaller unsigned type should have just worked if you convert back, like the old checked_cast code did, but callers of emit_intxxx were sign extending so the conversion back didn't work.

Calls like emit_int8() are passed arguments like 0xf0, but there were multiple (~940) that are passed with something that did some bitwise calculation that was promoted up to int.

Maybe the emit_intxx functions should take signed int, and then I could just check range in the checked_cast() without converting the argument to signed.  That might work and still provide the range checking that we want.

Yes, it should be max-1.  Thank you for looking at this.

Reply:
Changing checked_cast<> to not convert to a signed type for T1 leads to very puzzling compiler errors.  For these emit_int8 etc functions, maybe a straight cast is the best thing, and leaving the definition checked_cast<> alone.  I'd be happy to do that as I had in my first patch.

Reply:
I don't know what checked_cast<> should do for negative values when going between signed --> unsigned and for values greater than signed max for unsigned --> signed.  It seems like both should fail because the sign changes.